### PR TITLE
[GEOT-6158] include of xml schema does not work from within jar files

### PR DESCRIPTION
**Pull request by wumpz**
This solves the problem of using schema files with includes from jar files.

### CODE REVIEW

File: modules/extension/xsd/xsd-gml2/src/main/java/org/geotools/gml2/ApplicationSchemaXSD.java [hunk 1]
```diff
@@ -107,6 +107,12 @@ public String resolveSchemaLocation(XSDSchema schema, String uri, String locatio
                     if (locationFile.exists()) {
                         locationUri = locationFile.toURI().toString();
                     }
+
+                    if (locationUri == null
+                            && location != null
+                            && schemaLocationFolder.startsWith("jar:file:")) {
+                        locationUri = schemaLocationFolder + "/" + location;
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml2/src/main/java/org/geotools/gml2/ApplicationSchemaXSD.java]: Not mandatory, but I would add a comment, something like:
```
// handle schemas included in a JAR file
```

File: modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2002-2008, Open Source Geospatial Foundation (OSGeo)
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java]: Wrong date.

File: modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+package org.geotools.gml2;
+
+import static org.junit.Assert.*;
+
+import org.geotools.xsd.SchemaLocationResolver;
+import org.junit.Test;
+
+public class ApplicationSchemaXSDTest {
+
+    @Test
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java]: Please add a test describing what this class is testing.

File: modules/extension/xsd/xsd-gml3/src/main/java/org/geotools/gml3/ApplicationSchemaXSD.java [hunk 4]
```diff
@@ -105,6 +105,12 @@ public String resolveSchemaLocation(XSDSchema schema, String uri, String locatio
                     if (locationFile.exists()) {
                         locationUri = locationFile.toURI().toString();
                     }
+
+                    if (locationUri == null
+                            && location != null
+                            && schemaLocationFolder.startsWith("jar:file:")) {
+                        locationUri = schemaLocationFolder + "/" + location;
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml3/src/main/java/org/geotools/gml3/ApplicationSchemaXSD.java]: Same comment as above.

File: modules/extension/xsd/xsd-gml3/src/test/java/org/geotools/gml3/ApplicationSchemaXSDTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.geotools.gml3;
+
+import static org.junit.Assert.*;
+
+import org.geotools.xsd.SchemaLocationResolver;
+import org.junit.Test;
+
+public class ApplicationSchemaXSDTest {
+    @Test
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml3/src/test/java/org/geotools/gml3/ApplicationSchemaXSDTest.java]: Missing empty line, same comments as above :)

File: modules/extension/xsd/xsd-gml3/src/test/java/org/geotools/gml3/ApplicationSchemaXSDTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2002-2008, Open Source Geospatial Foundation (OSGeo)
```
Comment (nmco) [inline modules/extension/xsd/xsd-gml3/src/test/java/org/geotools/gml3/ApplicationSchemaXSDTest.java]: Same comment as above.

**Comment by nmco** [comment]
Review done, thank you for the contribution. By the way, you already provided [a code contribute license](http://docs.geotools.org/latest/developer/procedures/contribution_license.html) right ?

File: modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java (hunk 3 above)
Comment (wumpz) [reply modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java]: What do you mean? A comment?

**Comment by wumpz** [comment]
> Review done, thank you for the contribution. By the way, you already provided [a code contribute license](http://docs.geotools.org/latest/developer/procedures/contribution_license.html) right ?

No I haven't but I am on it. Why is this necessary? Never heard of OS software needed an agreement like that.

File: modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java (hunk 3 above)
Comment (nmco) [reply modules/extension/xsd/xsd-gml2/src/test/java/org/geotools/gml2/ApplicationSchemaXSDTest.java]: Yes a comment, thx.

**Comment by aaime** [comment]
Actually many open source projects do (mostly large ones). 
GeoTools and GeoServer both require you to sign and send a contribution agreement towards OSGeo the first time you change more than one non test file (like in this PR):
https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md

**Comment by aaime** [comment]
@wumpz have you by any chance sent the CLA?

**Comment by wumpz** [comment]
Did not yet get it back. Sorry.

**Comment by aaime** [comment]
@wumpz sorry I don't understand your comment... you mean you sent it to OSGeo, but got not reply?

**Comment by wumpz** [comment]
I gave it to my boss, to get it signed for the company but did not yet get it back. Sorry

**Comment by aaime** [comment]
Bummer, let's wait a bit more then :-)

**Comment by aaime** [comment]
Any news? :-)

**Comment by wumpz** [comment]
Finally, it's there. That was a long way to go ;). I sent it a minute ago. How long does it take to confirm?

**Comment by aaime** [comment]
It's all based on volunteer time, so, unpredictable.

**Comment by wumpz** [comment]
I got the confirmation, that our contribution agreement was approved. Therefore you can merge.

**Comment by aaime** [comment]
Confirmed and merged. Do you want to make backport pull requests too?

**Comment by wumpz** [comment]
For the last release version or more?

**Comment by aaime** [comment]
We are currently cutting releases out of 20.x and 21.x, backport to whatever of the two fits you best  (or both).

**Pull request merged**